src/Microsoft.Performance.SDK.Runtime/Discovery/AssemblyExtensionDiscovery.cs

```diff
@@ -203,70 +238,177 @@ public bool ProcessAssemblies(IEnumerable<string> directoryPaths)
                             var assembly = this.assemblyLoader.LoadAssembly(filePath);
                             if (!(assembly is null))
                             {
-                                allLoaded = allLoaded && ProcessAssembly(assembly);
-                            } else
+                                if (!ProcessAssembly(assembly, out var assemblyError))
+                                {
+                                    allLoaded = false;
+                                    Debug.Assert(assemblyError != null);
+                                    assemblyErrors.Add(assemblyError);
+                                }
+                            }
+                            else
                             {
+                                assemblyErrors.Add(
+                                    new ErrorInfo(ErrorCodes.AssemblyLoadFailed, "The assembly failed to load.")
+                                    {
+                                        Target = filePath,
+                                    });
                                 allLoaded = false;
                             }
                         }
                     }
+
+                    if (assemblyErrors.Count > 0)
+                    {
+                        directoryErrors.Add(
+                            new ErrorInfo(ErrorCodes.PluginRegistrationFailed, "Unable to load the plugin.")
```

Comment:
This is more of a conceptual point, but I do not think the `ProcessAssemblies` methods should be speaking in terms of plugins. These methods are just passed a bunch of assemblies to load up, and there could be multiple plugins (or even none!) inside of them. If anyone has authority to say a plugin load failed, it'd be the `PluginsLoader`, but even that may not be true because, again, it's just given directories.

I think we should just keep this (and the `PluginsLoader`'s) interface in terms of "the directory/directories you gave me did not load, here is the relevant reason why." If there's an error, whoever called the loader needs to figure out what that means in terms of plugins. In the future, having a `PluginsManager` that _knows_ a directory given to the `PluginsLoader` is a single plugin, then it'd have authority to use this error code.